File: src/main/java/io/myosotisdev/minestom/command/GameModeCommand.java
package io.myosotisdev.minestom.command;

import com.google.common.base.Strings;
import io.myosotisdev.minestom.Minestom;
import io.myosotisdev.minestom.permission.Permission;
import io.myosotisdev.utopianism.util.Namespaces;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.format.NamedTextColor;
import net.kyori.adventure.text.format.TextColor;
import net.minestom.server.command.CommandSender;
import net.minestom.server.command.ConsoleSender;
import net.minestom.server.command.builder.arguments.ArgumentEnum;
import net.minestom.server.command.builder.arguments.ArgumentString;
import net.minestom.server.command.builder.arguments.ArgumentType;
import net.minestom.server.command.builder.arguments.minecraft.ArgumentEntity;
import net.minestom.server.entity.GameMode;
import net.minestom.server.entity.Player;

public class GameModeCommand
        extends AbstractCommand
{
    public static final String                 Name    = "gamemode";
    public static final String[]               Aliases = new String[]
            {
                    "gm"
            };
    private             ArgumentEnum<GameMode> arg0;
    private             ArgumentEntity         arg1;

    public GameModeCommand()
    {
        super(null, Name, Aliases);

        this.arg0 = ArgumentType.Enum("gamemode", GameMode.class);
        this.arg1 = ArgumentType.Entity("player")
                                .onlyPlayers(true)
                                .singleEntity(true);

        addSyntax((sender, context) -> {
            if (!Minestom.checkPermission(sender, permission()))
            {
                sender.sendMessage(Component.text("You don't have enough permission to do that", NamedTextColor.RED));
                return;
            }

            if (sender instanceof ConsoleSender)
            {
                sender.sendMessage(Component.text("This command could only be used by player!", NamedTextColor.RED));
                return;
            }

            GameMode newMode = context.get(arg0);

            ((Player) sender).setGameMode(newMode);
            sender.sendMessage("Your gamemode has been set to " + newMode.name());
        }, arg0);
        addSyntax((sender, context) -> {
            if (!Minestom.checkPermission(sender, permission()))
            {
                sender.sendMessage(Component.text("You don't have enough permission to do that", NamedTextColor.RED));
                return;
            }

            Player player = context.get(arg1)
                                   .findFirstPlayer(null);
            if (player != null)
            {
                GameMode newMode = context.get(arg0);

                player.setGameMode(newMode);
                player.sendMessage("Your gamemode has been set to " + newMode.name());
            }
            else
                sender.sendMessage(Component.text("This command could only be used by player!", TextColor.color(255, 0, 0)));
        }, arg0, arg1);
    }

    void defaultExecutor(CommandSender sender, GameMode gameMode, Player player)
    {
        if (player == null)
        {
            if (sender instanceof ConsoleSender)
            {
                sender.sendMessage(Component.text("This command could only be used by player!", TextColor.color(255, 0, 0)));
                return;
            }
            else if (sender instanceof Player)
                player = ((Player) sender);
            else
            {
                sender.sendMessage("Unspecified sender type.");
                return;
            }
        }

        player.sendMessage("Your gamemode has been set to " + gameMode.name());
        player.setGameMode(gameMode);
    }

    @Override
    public Permission permission()
    {
        return Permission.ofString(Namespaces.Minecraft, "gamemode");
    }
}

File: src/main/kotlin/io/myosotisdev/minestom/command/GameModeCommand.kt
package io.myosotisdev.minestom.command

import io.myosotisdev.minestom.permission.Permission
import io.myosotisdev.minestom.permission.Permission.Companion.ofString
import io.myosotisdev.minestom.util.Components
import io.myosotisdev.utopianism.util.Namespaces
import net.minestom.server.command.builder.arguments.ArgumentType
import net.minestom.server.command.builder.arguments.ArgumentWord
import net.minestom.server.command.builder.arguments.minecraft.ArgumentEntity
import net.minestom.server.command.builder.arguments.number.ArgumentInteger
import net.minestom.server.command.builder.arguments.number.ArgumentNumber
import net.minestom.server.entity.GameMode
import net.minestom.server.entity.Player

class GameModeCommand : AbstractCommand(null, Name, *Aliases)
{
    companion object
    {
        const val Name = "gamemode"
        val Aliases = arrayOf(
                "gm"
        )
    }

    val arg0_mode: ArgumentWord = ArgumentType.Word("gamemode")
            .from(*GameMode.values()
                    .asList()
                    .map { gameMode ->
                        gameMode.toString()
                                .lowercase()
                    }
                    .toTypedArray())
    val arg0_modeInt: ArgumentNumber<Int> = ArgumentInteger("gamemode").between(0, 3);
    val arg1_player: ArgumentEntity = ArgumentType.Entity("player")
            .onlyPlayers(true)

    init
    {
        addSyntax({ sender, context ->
            val newMode = GameMode.valueOf(context.get(arg0_mode).uppercase())
            (sender as Player).gameMode = newMode
            sender.sendMessage("Your gamemode has been set to " + newMode.name)
        }, arg0_mode)

        addSyntax({ sender, context ->
            val player = context.get(arg1_player)
                    .findFirstPlayer(sender)
            if (player != null)
            {
                val newMode = GameMode.valueOf(context.get(arg0_mode).uppercase())
                player.gameMode = newMode
                player.sendMessage("Your gamemode has been set to " + newMode.name)
            }
            else sender.sendMessage(Components.fromLegacy("&cPlayer does not exist."))
        }, arg0_mode, arg1_player)

        addSyntax({ sender, context ->
            val newMode = GameMode.fromId(context.get(arg0_modeInt))
            (sender as Player).gameMode = newMode
            sender.sendMessage("Your gamemode has been set to " + newMode.name)
        }, arg0_modeInt)

        addSyntax({ sender, context ->
            val player = context.get(arg1_player)
                    .findFirstPlayer(sender)
            if (player != null)
            {
                val newMode = GameMode.fromId(context.get(arg0_modeInt))
                player.gameMode = newMode
                player.sendMessage("Your gamemode has been set to " + newMode.name)
            }
            else sender.sendMessage(Components.fromLegacy("&cPlayer does not exist."))
        }, arg0_modeInt, arg1_player)
    }

    override fun permission(): Permission?
    {
        return ofString(Namespaces.Minecraft, "gamemode")
    }
}